staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_interval.go

```diff
@@ -0,0 +1,245 @@
+	// index in the actual watch cache, however
+	// this should be used modulo the capacity
+	// of the cache whenever being used to index
+	// the circular buffer.
+	endIndex int
+
+	// indexer takes an index and returns the corresponding
+	// watchCacheEvent from an underlying watch cache.
+	indexer func(int) *watchCacheEvent
+
+	// intervalBuffer is a buffer of fixed size to hold the
+	// watchCacheEvents that this interval returns on a call to
+	// Next(). This exists mainly to reduce acquiring the lock
+	// on each invocation of Next().
+	intervalBuffer *watchCacheIntervalBuffer
+
+	// lock effectively guards the valid field.
+	lock sync.Locker
+
+	// valid indicates whether the cache interval
+	// being considered (b/w startIndex and endIndex)
+	// is still valid or not.
+	valid bool
+}
+
+type attrFunc func(runtime.Object) (labels.Set, fields.Set, error)
+
+func newCacheInterval(startIndex, endIndex int, locker sync.Locker) *watchCacheInterval {
+	return &watchCacheInterval{
+		startIndex:     startIndex,
+		endIndex:       endIndex,
+		intervalBuffer: &watchCacheIntervalBuffer{buffer: make([]*watchCacheEvent, bufferSize)},
+		lock:           locker,
+	}
+}
+
+// When dealing with the case of resource version = 0, the events returned by Next()
+// need to be events from a List() done on the underlying store of the watch cache.
+func newCacheIntervalFromStore(store cache.Indexer, getAttrsFunc attrFunc, locker sync.Locker) (*watchCacheInterval, error) {
+	ciBuffer := &watchCacheIntervalBuffer{}
+	allItems := store.List()
+	ciBuffer.buffer = make([]*watchCacheEvent, len(allItems))
+	for i, item := range allItems {
+		elem, ok := item.(*storeElement)
+		if !ok {
+			return nil, fmt.Errorf("not a storeElement: %v", elem)
+		}
+		objLabels, objFields, err := getAttrsFunc(elem.Object)
+		if err != nil {
+			return nil, err
+		}
+		ciBuffer.buffer[i] = &watchCacheEvent{
+			Type:            watch.Added,
+			Object:          elem.Object,
+			ObjLabels:       objLabels,
+			ObjFields:       objFields,
+			Key:             elem.Key,
+			ResourceVersion: 0,
+		}
+	}
+	ci := &watchCacheInterval{intervalBuffer: ciBuffer, lock: locker}
+	// simulate that we already have all the events we're looking for.
+	ci.startIndex = ci.endIndex
+
+	return ci, nil
+}
+
+// An indexer is registered when dealing with cases of resource version > 0, i.e. we
+// want to read from the underlying watch cache and not the underlying store.
+func (wci *watchCacheInterval) registerIndexerFunc(indexer func(int) *watchCacheEvent) {
+	wci.indexer = indexer
+}
+
+// Next returns the next item in the cache interval provided the cache
+// interval is still valid. An error is returned if the interval is
+// invalidated.
+func (wci *watchCacheInterval) Next() (*watchCacheEvent, bool, error) {
+	// if there are items in the buffer to return, return from
+	// the buffer.
+	if !wci.intervalBuffer.isIntervalBufferEmpty() {
+		return wci.intervalBuffer.getNextEvent(), true, nil
```

Comment:
Let's merget those into a single function:

event, exists := wci.intervalBuffer.next()
// next() returns (event, true) if it has event in buffer, otherwise returns (nil, false)
if exists {
  return event, exists, nil
}
...